api/src/main/java/org/apache/iceberg/expressions/BoundReference.java

```diff
@@ -26,10 +26,12 @@
 public class BoundReference<T> implements BoundTerm<T>, Reference<T> {
   private final Types.NestedField field;
   private final Accessor<StructLike> accessor;
+  private final String name;
 
-  BoundReference(Types.NestedField field, Accessor<StructLike> accessor) {
+  BoundReference(Types.NestedField field, Accessor<StructLike> accessor, String name) {
     this.field = field;
     this.accessor = accessor;
+    this.name = name;
```

Comment:
Tracking the field name was needed to handle bound expressions. This is also a good thing to track for other purposes (e.g., easier to read log messages with bound expressions).